tket/src/Circuit/SubcircuitFinder.cpp

```diff
@@ -0,0 +1,285 @@
+    BGL_FORALL_VERTICES(v, circ_->dag, DAG) {
+      if (criterion(circ_->get_Op_ptr_from_Vertex(v))) {
+        VertexVec preds = circ_->get_predecessors(v);
+        VertexVec succs = circ_->get_successors(v);
+        subcircuit_infos.push_back(
+            {{v}, {preds.begin(), preds.end()}, {succs.begin(), succs.end()}});
+      }
+    }
+    while (true) {
+      // If possible, find a mergeable pair of subcircuits and merge them.
+      std::optional<std::pair<std::size_t, std::size_t>> indices =
+          find_mergeable_pair(subcircuit_infos);
+      if (!indices.has_value()) {
+        break;
+      } else {
+        std::size_t i0 = indices->first;
+        std::size_t i1 = indices->second;
+        TKET_ASSERT(i0 < i1);
+        subcircuit_info_t subcircuit_info0 = subcircuit_infos[i0];
+        subcircuit_info_t subcircuit_info1 = subcircuit_infos[i1];
+        // Must erase in this order, since i0 < i1:
+        subcircuit_infos.erase(subcircuit_infos.begin() + i1);
+        subcircuit_infos.erase(subcircuit_infos.begin() + i0);
+        subcircuit_infos.push_back(
+            convex_union(subcircuit_info0, subcircuit_info1));
+      }
+    }
+    std::vector<VertexSet> subcircuits;
+    for (const subcircuit_info_t &subcircuit_info : subcircuit_infos) {
+      subcircuits.push_back(subcircuit_info.verts);
+    }
+    return subcircuits;
+  }
+
+ private:
+  // Test whether the union of two disjoint connected convex subcircuits is
+  // convex.
+  bool union_is_convex(
+      const subcircuit_info_t &subcircuit_info0,
+      const subcircuit_info_t &subcircuit_info1) {
+    const VertexSet &preds0 = subcircuit_info0.preds;
+    const VertexSet &succs0 = subcircuit_info0.succs;
+    const VertexSet &preds1 = subcircuit_info1.preds;
+    const VertexSet &succs1 = subcircuit_info0.succs;
+    for (const Vertex &v0 : succs0) {
+      for (const Vertex &v1 : preds1) {
+        if (order_relations_.contains({v0, v1})) {
+          return false;
+        }
+      }
+    }
+    for (const Vertex &v1 : succs1) {
+      for (const Vertex &v0 : preds0) {
+        if (order_relations_.contains({v1, v0})) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  // Given a vector of disjoint connected convex subcircuits, look for a pair
+  // whose union is connected and convex, and return the indices of such a pair
+  // if it exists.
+  std::optional<std::pair<std::size_t, std::size_t>> find_mergeable_pair(
+      const std::vector<subcircuit_info_t> &subcircuit_infos) {
+    std::size_t n_subcircuits = subcircuit_infos.size();
+    for (std::size_t i0 = 0; i0 < n_subcircuits; i0++) {
+      for (std::size_t i1 = i0 + 1; i1 < n_subcircuits; i1++) {
+        const subcircuit_info_t &subcircuit_info0 = subcircuit_infos[i0];
+        const subcircuit_info_t &subcircuit_info1 = subcircuit_infos[i1];
+        if (union_is_connected(subcircuit_info0, subcircuit_info1) &&
```

Comment:
> Is there a way to stop the inner loop early?

It will already stop early as soon as it finds a good pair -- or perhaps I misunderstand the question.

> Is it possible to only consider _neighbouring_ sets?

That is what `union_is_connected()` is meant to be checking. (It's intentionally put first so that we can skip the more involved convexity check if it's not connected.)

Reply:
Sorry, I wasn't very clear. Suppose we have a one qubit circuit:
`Circuit(1).X(0).T(0).H(0).T(0).Z(0).T(0).Y(0)`
if we find that X and H are not connected then we don't have to test (X, Z) or (X, Y).

Reply:
Hmm, true but in general we have no guarantee that if x < y < z in the list of subcircuits and x and y are not connected then x and z are not connected. Moreover I don't think it's possible to guarantee such an order: consider the circuit
```python
Circuit(2).CX(0,1).H(0).H(1).CX(0,1)
```
where each gate is a singleton subcircuit. There is no ordering of these four subcircuits with that property.

Reply:
I was thinking that if the subcircuits are in topological order then we can close up connection points (i.e. qubits) as we examine subcircuits in the inner loop.

If subcircuits c0 and c1 are not connected, but they share qubits q_n, q_m then any subsequent subcircuit c_i, where i>1,  is not connected to c0 if it has qubits q_n, q_m.

Reply:
Yes, that could be an optimization. Maintaining topological order could be tricky. When we merge two subcircuits, deciding where to place the merged circuit in the list would involve some extra checks. We'd also need to keep track of actual `UnitID`s associated with the edges.

Maybe this could be deferred to a later task to see if it significantly improves performance?

Reply:
Right, sounds more tricky than I initially thought.